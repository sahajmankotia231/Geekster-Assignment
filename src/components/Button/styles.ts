import styled from "styled-components";

import {
  WHITE,
  BLUE,
  LIGHT_BLUE,
  LIGHT_BLUE_BORDER,
  RED,
} from "constants/colors";
import { BUTTON_RADIUS } from "constants/styles";

const themeMap = Object.freeze({
  primary: Object.freeze({
    backgroundColor: BLUE,
    borderColor: BLUE,
    color: WHITE,
    outlineColor: BLUE,
  }),
  secondary: Object.freeze({
    backgroundColor: LIGHT_BLUE,
    borderColor: LIGHT_BLUE_BORDER,
    color: BLUE,
    outlineColor: BLUE,
  }),
  danger: Object.freeze({
    backgroundColor: RED,
    borderColor: RED,
    color: WHITE,
    outlineColor: RED,
  }),
});

const getBackgroundColor = (outline: boolean, theme: string): string => {
  return outline ? "transparent" : themeMap[theme].backgroundColor;
};

const getBorder = (outline: boolean, theme: string): string => {
  return outline
    ? "1px solid transparent"
    : `1px solid ${themeMap[theme].borderColor}`;
};

const getColor = (outline: boolean, theme: string): string => {
  return outline ? themeMap[theme].outlineColor : themeMap[theme].color;
};

export const StyledButton = styled.button`
  width: ${({ block }) => (block ? "100%" : "auto")};
  min-width: 40px;
  height: 40px;
  align-items: center;
  background-color: ${({ outline, theme }) =>
    getBackgroundColor(outline, theme)};
  border: ${({ outline, theme }) => getBorder(outline, theme)};
  border-radius: ${BUTTON_RADIUS}px;
  color: ${({ outline, theme }) => getColor(outline, theme)};
  cursor: pointer;
  display: ${({ block }) => (block ? "flex" : "inline-flex")};
  font-size: 16px;
  font-weight: 700;
  justify-content: center;
  line-height: 20px;

  svg {
    margin: 0 4px;
  }

  &:disabled {
    opacity: 0.4;
  }
`;
