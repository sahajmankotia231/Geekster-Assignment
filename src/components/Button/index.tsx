import React, { ReactElement, FunctionComponent } from "react";

import { StyledButton } from "./styles";

interface ButtonProps {
  theme?: string;
  onClick: () => void;
  outline?: boolean;
  block?: boolean;
  dataTestId?: string;
  disabled?: boolean;
}

const Button: FunctionComponent<ButtonProps> = ({
  theme = "primary",
  children,
  onClick,
  outline = false,
  block = true,
  dataTestId = "button",
  disabled = false,
}): ReactElement => {
  return (
    <StyledButton
      onClick={onClick}
      theme={theme}
      outline={outline}
      block={block}
      data-testid={dataTestId}
      disabled={disabled}
    >
      {children}
    </StyledButton>
  );
};

export default Button;
