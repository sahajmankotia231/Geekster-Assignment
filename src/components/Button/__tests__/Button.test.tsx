import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";

import Button from "..";

describe("<Button />", () => {
  it("should render properly", () => {
    const { container } = render(
      <Button onClick={() => {}}>Button content</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should trigger onClick prop", () => {
    const mockedOnClick = jest.fn();
    const buttonText = "Button Content";

    render(<Button onClick={mockedOnClick}>{buttonText}</Button>);

    fireEvent.click(screen.queryByText(buttonText));

    expect(mockedOnClick).toHaveBeenCalled();
  });

  it("could be disable", () => {
    const buttonText = "Button Content";

    const { container } = render(
      <Button onClick={() => {}} disabled>
        {buttonText}
      </Button>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
