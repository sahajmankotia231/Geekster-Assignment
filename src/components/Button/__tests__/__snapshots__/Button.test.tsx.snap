// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> could be disable 1`] = `
<button
  class="sc-bdvvaa tPxtg"
  data-testid="button"
  disabled=""
>
  Button Content
</button>
`;

exports[`<Button /> should render properly 1`] = `
<button
  class="sc-bdvvaa tPxtg"
  data-testid="button"
>
  Button content
</button>
`;
